### PR TITLE
Network sync refactoring (part 6)

### PR DESCRIPTION
**Pull request by nazar-pc**
Yet another one related to https://github.com/paritytech/substrate/issues/8686

This time I noticed that there are a few traits in the codebase that somewhat match `NetworkService` and are implemented for it. Also `NetworkService` itself has way too many methods for all kinds of things that are not used in most cases.

What I did then is sliced `NetworkService` into many logical traits and made interfaces generic in crates that previously depended this specific implementation.

Next PR will remove `sc-network` from dependencies of most crates, making Substrate a framework with pluggable networking, not just consensus. API provided by new traits is the same as before and may need to be generalized/cleaned up in the future, but first decoupling of the rest of Substrate from `sc-network` is needed.

Reviewing commit after commit is recommended. There is no business logic changes here, just moving things around a bit. PR is big, I can cut it in half at any commit if it helps, but all commits are of the same type here. Hopefully trait collection and naming makes sense.

Polkadot companion: https://github.com/paritytech/polkadot/pull/5841
Cumulus companion: https://github.com/paritytech/cumulus/pull/1486

Polkadot address: 1vSxzbyz2cJREAuVWjhXUT1ds8vBzoxn2w4asNpusQKwjJd

### CODE REVIEW

**Comment by nazar-pc** [comment]
Fixed documentation references

**Comment by nazar-pc** [comment]
@TriplEight I guess you already noticed this:
> Github API says polkadot#5841 is not mergeable

It isn't indeed because CI didn't pass there, but it is physically mergeable (no conflicts against `master`), so should be allowed to run here as a companion check I think.

**Comment by nazar-pc** [comment]
Hey @bkchr and @arkpar I'd like to get your feedback on this one

**Comment by arkpar** [comment]
Looks good here.

**Review by arkpar — APPROVED** [review]

File: client/service/src/builder.rs [hunk 1]
```diff
@@ -319,6 +322,30 @@ where
 	)
 }
 
+pub trait SpawnTaskNetwork<Block: BlockT>:
```
Comment (bkchr) [inline client/service/src/builder.rs:325]: Missing docs

File: client/network/common/src/protocol.rs [hunk 2]
```diff
@@ -0,0 +1 @@
+pub mod event;
```
Comment (bkchr) [inline client/network/common/src/protocol.rs:1]: Missing license header.

File: client/network/common/src/service.rs [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+use sc_peerset::ReputationChange;
+pub use signature::Signature;
+use sp_runtime::traits::{Block as BlockT, NumberFor};
+use std::{borrow::Cow, collections::HashSet, future::Future, pin::Pin, sync::Arc};
+
+mod signature;
+
+/// Signer with network identity
+pub trait NetworkSigner {
+	/// Signs the message with the `KeyPair` that defined the local `PeerId`.
```
Comment (bkchr) [inline client/network/common/src/service.rs:38]:
```suggestion
	/// Signs the message with the `KeyPair` that defines the local [`PeerId`].
```

File: client/network/common/src/service.rs [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+	T: ?Sized,
+	T: NetworkSigner,
+{
+	fn sign_with_local_identity(&self, msg: impl AsRef<[u8]>) -> Result<Signature, SigningError> {
+		T::sign_with_local_identity(self, msg)
+	}
+}
+
+/// Get value network provider.
+pub trait NetworkKVProvider {
```
Comment (bkchr) [inline client/network/common/src/service.rs:53]:
```suggestion
/// Provides access to the networking DHT.
pub trait NetworkDHTProvider {
```

File: client/network/common/src/service.rs [hunk 5]
```diff
@@ -0,0 +1,660 @@
...
+	/// Set authorized peers.
+	///
+	/// Need a better solution to manage authorized peers, but now just use reserved peers for
+	/// prototyping.
+	fn set_authorized_peers(&self, peers: HashSet<PeerId>);
+
+	/// Set authorized_only flag.
+	///
+	/// Need a better solution to decide authorized_only, but now just use reserved_only flag for
+	/// prototyping.
```
Comment (bkchr) [inline client/network/common/src/service.rs:156]: Are these comments copied?

File: client/network/common/src/service.rs [hunk 6]
```diff
@@ -0,0 +1,660 @@
...
+	/// Returns an `Err` if one of the given addresses is invalid or contains an
+	/// invalid peer ID (which includes the local peer ID).
+	fn add_peers_to_reserved_set(
+		&self,
+		protocol: Cow<'static, str>,
+		peers: HashSet<Multiaddr>,
+	) -> Result<(), String>;
+
+	/// Remove peers from a peer set.
+	fn remove_peers_from_reserved_set(&self, protocol: Cow<'static, str>, peers: Vec<PeerId>);
```
Comment (bkchr) [inline client/network/common/src/service.rs:228]: Most of these functions are basically duplicated, just with some small differences in their behavior. I think, merging them and controlling the exact behavior with an enum as arguments sounds better?

File: client/network/common/src/service.rs [hunk 7]
```diff
@@ -0,0 +1,660 @@
...
+		peers: HashSet<Multiaddr>,
+	) -> Result<(), String>;
+
+	/// Remove peers from a peer set.
+	///
+	/// If we currently have an open substream with this peer, it will soon be closed.
+	fn remove_from_peers_set(&self, protocol: Cow<'static, str>, peers: Vec<PeerId>);
+
+	/// Returns the number of peers we're connected to.
+	fn num_connected(&self) -> usize;
```
Comment (bkchr) [inline client/network/common/src/service.rs:251]:
```suggestion
	/// Returns the number of peers in the sync peer set we're connected to.
	fn sync_num_connected(&self) -> usize;
```

File: client/network/common/src/service.rs [hunk 8]
```diff
@@ -0,0 +1,660 @@
...
+		T::local_peer_id(self)
+	}
+}
+
+/// Reserved slot in the notifications buffer, ready to accept data.
+pub trait NotificationSenderReady {
+	/// Consumes this slots reservation and actually queues the notification.
+	///
+	/// NOTE: Traits can't consume itself, but calling this method second time will return an error.
+	fn send(&mut self, notification: Vec<u8>) -> Result<(), NotificationSenderError>;
```
Comment (bkchr) [inline client/network/common/src/service.rs:385]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2dedb5178382fd9b26480a870098834d

:thinking:

File: client/network/common/src/service.rs [hunk 9]
```diff
@@ -0,0 +1,660 @@
...
+}
+
+/// A `NotificationSender` allows for sending notifications to a peer with a chosen protocol.
+#[async_trait::async_trait]
+pub trait NotificationSender {
+	/// Returns a future that resolves when the `NotificationSender` is ready to send a
+	/// notification.
+	async fn ready(&self)
+		-> Result<Box<dyn NotificationSenderReady + '_>, NotificationSenderError>;
```
Comment (bkchr) [inline client/network/common/src/service.rs:394]: Why not have an associated type as return value?

**Review by bkchr — APPROVED** [review]
I'm approving this as it looks good. However, we should stop here. There is no need to have Substrate being completely opaque to the underlying network. This mainly just introduces complexity that we don't really need.

The task should really be to make networking agnostic that there exists a syncing protocol. It should not need to care. The networking crate should only provide ways to register protocols and ensure to drive these protocols. One of these protocols is then syncing.

**Comment by bkchr** [comment]
@nazar-pc and also please provide your polkadot address for a tip :)

File: client/network/common/src/service.rs (hunk 5 above)
Comment (nazar-pc) [reply client/network/common/src/service.rs:156]: Yes, all of them

File: client/network/common/src/service.rs (hunk 6 above)
Comment (nazar-pc) [reply client/network/common/src/service.rs:228]: Not sure, having separate names is more convenient IMO. The primary goal in this PR is to not change API if at all possible, just moving things around.

File: client/network/common/src/service.rs (hunk 8 above)
Comment (nazar-pc) [reply client/network/common/src/service.rs:385]: You can define it, but can't call it later: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f01675569f20b3c7c6bc96cc0e3f0565

File: client/network/common/src/service.rs (hunk 9 above)
Comment (nazar-pc) [reply client/network/common/src/service.rs:394]: I believe it either doesn't work at all or becomes tricky with trait object when we want to use it.

**Comment by nazar-pc** [comment]
> I'm approving this as it looks good. However, we should stop here. There is no need to have Substrate being completely opaque to the underlying network. This mainly just introduces complexity that we don't really need.

I have just one small upcoming PR where explicit `sc-network` dependency is replaced in a few places with interfaces from `sc-network-common`, but no other changes planned as it is already abstracted away completely and basically everywhere (this have been a bit of detour I agree, but we did need this in our protocol). Substrate aims to be modular, sorry if I went too far here :see_no_evil: 

> The task should really be to make networking agnostic that there exists a syncing protocol. It should not need to care. The networking crate should only provide ways to register protocols and ensure to drive these protocols. One of these protocols is then syncing.

Understood.

> and also please provide your polkadot address for a tip :)

:heavy_check_mark:

**Comment by bkchr** [comment]
> I have just one small upcoming PR where explicit `sc-network` dependency is replaced in a few places with interfaces from `sc-network-common`, but no other changes planned as it is already abstracted away completely and basically everywhere (this have been a bit of detour I agree, but we did need this in our protocol). Substrate aims to be modular, sorry if I went too far here see_no_evil

This should be fine, especially if you have some need for this :) Just wanted to highlight that we should stop here :grimacing:

File: client/network/common/src/service.rs (hunk 9 above)
Comment (bkchr) [reply client/network/common/src/service.rs:394]: Then we don't need any trait object? And then we also can make the function consume the object. Sounds like a win win to me :D

Comment (bkchr) [reply client/network/common/src/service.rs:394]: Hmm okay

**Comment by bkchr** [comment]
bot merge

**Pull request merged**